src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/TrieBuilder.cs

```diff
@@ -0,0 +1,509 @@
+                    {
+                        newNode = fAdd(ref this, nodes[i], state, out bool canContinueInner);
+                        if (!visited.Add(newNode))
+                        {
+                            continue;
+                        }
+                        // See comments in handling RegexNodeKind.Alternate below, to understand how it works.
+                        if (canContinueInner)
+                        {
+                            canContinue = true;
+                            result.Add(newNode);
+                        }
+                        else
+                        {
+                            _nodes[newNode].IsMatch = true;
+                        }
+                    }
+                    return new NodeCollection(result);
+            }
+        }
+
+        /// <summary>
+        /// Adds a character after each node of the given collection.
+        /// </summary>
+        /// <param name="nodes">The collection of nodes</param>
+        /// <param name="c">The character to add.</param>
+        /// <param name="canContinue">Returns whether the traversal algorithm can look past the character.</param>
+        /// <returns>The collection of node indices that each node of
+        /// <paramref name="nodes"/> leads to, at <paramref name="c"/>.</returns>
+        private NodeCollection Add(NodeCollection nodes, char c, out bool canContinue)
+        {
+            return AddOneToOneHelper(nodes, c, Add, out canContinue);
+        }
+
+        /// <summary>
+        /// Adds a character after the given node.
+        /// </summary>
+        /// <param name="builder">The <see cref="TrieBuilder"/> <paramref name="nodeIndex"/> belongs to.</param>
+        /// <param name="nodeIndex">The node's index.</param>
+        /// <param name="c">The character to add.</param>
+        /// <param name="canContinue">Returns whether the traversal algorithm can look past the character.</param>
+        /// <returns>The index of the node that <paramref name="nodeIndex"/>
+        /// leads to, at <paramref name="c"/>.</returns>
+        private static int Add(ref TrieBuilder builder, int nodeIndex, char c, out bool canContinue)
+        {
+            List<TrieNode> nodes = builder._nodes;
+            TrieNode node = nodes[nodeIndex];
+            if (builder.IsNodeLimitReached || node.Depth >= DepthLimit)
+            {
+                canContinue = false;
+                return nodeIndex;
+            }
+
+            if (!node.Children.TryGetValue(c, out int nextNodeIndex))
+            {
+                nextNodeIndex = nodes.Count;
+
+                TrieNode newNode = new TrieNode()
+                {
+                    Parent = nodeIndex,
+                    AccessingCharacter = c,
+                    Depth = node.Depth + 1,
+#if DEBUG
+                    Path = node.Path + c
+#endif
+                };
+
+                node.Children.Add(c, nextNodeIndex);
```

Comment:
I had the idea that if the regex is case-insensitive, it would add a transition via all case-equivalent characters to the same new node, but I need a `CutureInfo` to use `RegexCaseEquivalences`, and it seems to me that at this point, the case-insensitivity of the regex has been stripped away. 🤔

Reply:
There's no difference between "(?i)ab" and "[Aa][Bb]".  If the goal is to apply a transition for all case-equivalent characters, you'd just do so for all members of the set.

Reply:
An idea I had is to merge equivalent states after constructing the trie links. It could greatly reduce the state count on case-insensitive regexes, but it needs some thought on how to do, would happen after taking the trie node limit into account, and might be expensive. I'd prefer to explore this in the future.